protocols/virtualfund/virtualfund_single_hop_test.go

```diff
@@ -0,0 +1,304 @@
+	}
+}
+
+func collectPeerSignaturesOnSetupState(V *channel.SingleHopVirtualChannel, myRole uint, prefund bool) {
+	var state state.State
+	if prefund {
+		state = V.PreFundState()
+	} else {
+		state = V.PostFundState()
+	}
+
+	if myRole != alice.role {
+		aliceSig, _ := state.Sign(alice.privateKey)
+		V.AddStateWithSignature(state, aliceSig)
+	}
+	if myRole != p1.role {
+		p1Sig, _ := state.Sign(p1.privateKey)
+		V.AddStateWithSignature(state, p1Sig)
+	}
+	if myRole != bob.role {
+		bobSig, _ := state.Sign(bob.privateKey)
+		V.AddStateWithSignature(state, bobSig)
+	}
+}
+
+func TestCrankAsAlice(t *testing.T) {
+	my := alice
+	td := newTestData()
+	vPreFund := td.vPreFund
+	ledgers := td.ledgers
+	var s, _ = constructFromState(false, vPreFund, my.address, ledgers[my.destination].left, ledgers[my.destination].right) // todo: #420 deprecate TwoPartyLedgers
+	// Assert that cranking an unapproved objective returns an error
+	_, _, _, err := s.Crank(&my.privateKey)
+	assert(t, err != nil, `Expected error when cranking unapproved objective, but got nil`)
+
+	// Approve the objective, so that the rest of the test cases can run.
+	o := s.Approve().(*Objective)
+
+	// To test the finite state progression, we are going to progressively mutate o
+	// And then crank it to see which "pause point" (WaitingFor) we end up at.
+	// NOTE: Because crank returns a protocools.Objective interface, after each crank we
+	// need to remember to convert the result back to a virtualfund.Objective struct
+
+	// Initial Crank
+	oObj, got, waitingFor, err := o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+
+	expectedSignedState := state.NewSignedState(o.V.PreFundState())
+	mySig, _ := o.V.PreFundState().Sign(my.privateKey)
+	_ = expectedSignedState.AddSignature(mySig)
+
+	ok(t, err)
+	equals(t, waitingFor, WaitingForCompletePrefund)
+	assertStateSentTo(t, got, expectedSignedState, bob)
+	assertStateSentTo(t, got, expectedSignedState, p1)
+
+	// Manually progress the extended state by collecting prefund signatures
+	collectPeerSignaturesOnSetupState(o.V, my.role, true)
+
+	// Cranking should move us to the next waiting point, update the ledger channel, and alter the extended state to reflect that
+	// TODO: Check that ledger channel is updated as expected
+	oObj, got, waitingFor, err = o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+
+	p := consensus_channel.NewAddProposal(o.ToMyRight.Channel.Id, 2, o.ToMyRight.getExpectedGuarantee(), big.NewInt(6))
+	sp := consensus_channel.SignedProposal{Proposal: p}
+	assertProposalSent(t, got, sp, p1)
+	ok(t, err)
+	equals(t, waitingFor, WaitingForCompleteFunding)
+
+	// Check idempotency
+	emptySideEffects := protocols.SideEffects{}
+	oObj, got, waitingFor, err = o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+	ok(t, err)
+	equals(t, got, emptySideEffects)
+	equals(t, waitingFor, WaitingForCompleteFunding)
+}
+
+// Copied from https://github.com/benbjohnson/testing
+
+// assert fails the test if the condition is false.
```

Comment:
Do we expect to use these in other tests? Should we move them somewhere central and add some better godoc?

Reply:
That is definitely a good idea!

I've created https://github.com/statechannels/go-nitro/issues/517 -- it could be useful to use these assertions in some other tests to make them more readable.